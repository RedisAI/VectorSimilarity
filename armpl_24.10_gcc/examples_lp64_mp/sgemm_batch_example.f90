!   SGEMM Example Program Text
!
!   Arm Performance Libraries version 24.10
!   SPDX-FileCopyrightText: Copyright 2015-2024 Arm Limited and/or its affiliates
!
    Program main
!      .. Use Statements ..
      Use armpl_library, Only: sgemm_batch
!      .. Implicit NONE Statement ..
      Implicit None
!      .. Parameters ..
      Integer, Parameter               :: nmax = 250
!      .. Local scalars ..
      Integer                          :: i, j, ii, jj
      Integer                          :: group_count, total_batch_count
!      .. Local Arrays ..
      Integer, Allocatable, Target     :: m(:), n(:), k(:)
      Integer, Allocatable             :: lda(:), ldb(:), ldc(:), &
                                          group_size(:)
      Integer, Pointer                 :: ma, na, mb, nb
      Real, Allocatable                :: alpha(:), beta(:)
      Real, Allocatable                :: a(:, :), b(:, :), c(:, :)
!      .. Arrays for transpose options ..
      Character *1, Allocatable        :: transa(:), transb(:)
!      .. Arrays for pointers to matrices ..
      Integer *8, Allocatable          :: ap(:), bp(:), cp(:)
      Integer                          :: aptr, bptr, cptr, ki
!      .. Executable statements ..

       Write (*, Fmt=99995)                                            &
           'ARMPL example for SGEMM_BATCH'
       Write (*, Fmt=99995)                                            &
           '-------------------------------------------------------'
      Write (*, *) ''

      group_count = 3
      Allocate (m(group_count), n(group_count), k(group_count), &
        lda(group_count), ldb(group_count), ldc(group_count), &
        group_size(group_count), alpha(group_count), beta(group_count), &
        a(nmax,nmax), b(nmax,nmax), c(nmax,nmax), transa(group_count), &
        transb(group_count))
!      Set values to matrices a, b and c
      Do j = 1, nmax
        Do i = 1, nmax
          a(i, j) = i + j
          b(i, j) = i - j
          c(i, j) = 0
        End Do
      End Do

!      Set array with maxtrix A's transpose options
      transa(1) = 'N'
      transa(2) = 'T'
      transa(3) = 'N'

!      Set array with matrix B's transpose options
      transb(1) = 'N'
      transb(2) = 'T'
      transb(3) = 'N'

!      Set sizes for matrices
!      Set values for m array
      m(1) = 2
      m(2) = 2
      m(3) = 3

!      Set values for k array
      k(1) = 3
      k(2) = 2
      k(3) = 3

!      Set values for n array
      n(1) = 2
      n(2) = 3
      n(3) = 2

!      Set values for alpha array
      alpha(1) = 1.0
      alpha(2) = 1.0
      alpha(3) = 1.0

!      Set values for beta array
      beta(1) = 0.0
      beta(2) = 0.0
      beta(3) = 0.0

!      Set values for lda array
      lda(1) = nmax
      lda(2) = nmax
      lda(3) = nmax

!      Set values for ldb array
      ldb(1) = nmax
      ldb(2) = nmax
      ldb(3) = nmax

!      Set values for ldc array
      ldc(1) = nmax
      ldc(2) = nmax
      ldc(3) = nmax

!      Set values for group size array
      group_size(1) = 2
      group_size(2) = 3
      group_size(3) = 4

      total_batch_count = 0
      Do i = 1, group_count
        total_batch_count = total_batch_count + group_size(i)
      End Do

      Allocate (ap(total_batch_count), bp(total_batch_count), &
        cp(total_batch_count))

!      Set pointers for A matrices
!      Pointers to group 1
!      - trans opt = N. m = 2
      ap(1) = loc(a(1,1))
      ap(2) = loc(a(3,1))
!      Pointers to group 2
!      - trans opt = T, k = 2
      ap(3) = loc(a(5,1))
      ap(4) = loc(a(7,1))
      ap(5) = loc(a(9,1))
!      Pointers to group 3
!      - trans opt = N, m = 3
      ap(6) = loc(a(11,1))
      ap(7) = loc(a(14,1))
      ap(8) = loc(a(17,1))
      ap(9) = loc(a(20,1))

!      Set pointers to B matrices
!      Pointers to group 1
!      - trans opt = N, k = 3
      bp(1) = loc(b(1,1))
      bp(2) = loc(b(4,1))
!      Pointers to group 2
!      - trans opt = T, n = 3
      bp(3) = loc(b(7,1))
      bp(4) = loc(b(10,1))
      bp(5) = loc(b(13,1))
!      Pointers to group 3
!      - trans opt = N, k = 3
      bp(6) = loc(b(16,1))
      bp(7) = loc(b(19,1))
      bp(8) = loc(b(22,1))
      bp(9) = loc(b(25,1))

!      Set pointers to C matrices
!      Pointers to group 1
!      m = 2
      cp(1) = loc(c(1,1))
      cp(2) = loc(c(3,1))
!      Pointer to group 2
!      m = 2
      cp(3) = loc(c(5,1))
      cp(4) = loc(c(7,1))
      cp(5) = loc(c(9,1))
!      Pointers to group 3
!      m = 3
      cp(6) = loc(c(11,1))
      cp(7) = loc(c(14,1))
      cp(8) = loc(c(17,1))
      cp(9) = loc(c(20,1))

!      Call the function
      Call sgemm_batch(transa, transb, m, n, k, alpha, ap, lda, bp, ldb, beta, &
        cp, ldc, group_count, group_size)

!
!      Print all the data
!
!      Pointer to matrices
      aptr = 0
      bptr = 0
      cptr = 0
      ki = 1

      Do i = 1, group_count

        If (transa(i)=='N') Then
          ma => m(i)
          na => k(i)
        Else
          ma => k(i)
          na => m(i)
        End If

        If (transb(i)=='N') Then
          mb => k(i)
          nb => n(i)
        Else
          mb => n(i)
          nb => k(i)
        End If

!        Print each entry of each group
        Write (*, Fmt=99999) i
        Do j = 1, group_size(i)
!           Print matrix A
          Write (*, Fmt=99995) '**** Matrix A'
          Write (*, Fmt=99996) transa(i)
          Do ii = 1, ma
            Write (*, Fmt=99997)(a(aptr+ii,jj), jj=1, na)
          End Do
          Write (*, *) ''
!           Print matrix B
          Write (*, Fmt=99995) '**** Matrix B'
          Write (*, Fmt=99996) transb(i)
          Do ii = 1, mb
            Write (*, Fmt=99997)(b(bptr+ii,jj), jj=1, nb)
          End Do
          Write (*, *) ''
          Write (*, Fmt=99998) 'alpha : ', alpha(i)
          Write (*, Fmt=99998) 'beta  : ', beta(i)
          Write (*, *) ''
!           Print matrix C
          Write (*, Fmt=99995) '**** Matrix C'
          Do ii = 1, m(i)
            Write (*, Fmt=99997)(c(cptr+ii,jj), jj=1, n(i))
          End Do
          Write (*, *) ''

!           Update pointer based on transpose option
          If (transa(i)=='N') Then
            aptr = aptr + m(i)
          Else
            aptr = aptr + k(i)
          End If

          If (transb(i)=='N') Then
            bptr = bptr + k(i)
          Else
            bptr = bptr + n(i)
          End If

          cptr = cptr + m(i)

          ki = ki + 1

          Write (*, Fmt=99995) '=============end entry'
        End Do
        Write (*, Fmt=99995) '=============end group============='

      End Do

      Deallocate (m, n, k, lda, ldb, ldc, group_size, alpha, beta, a, b, c, &
        ap, bp, cp, transa, transb)
99999 Format ('******** Group: ', I2)
99998 Format ('** ', A, F7.3)
99997 Format (1X, 1P, 7E13.3)
99996 Format ('** Trans option: ', A1)
99995 Format (A)
    End Program
