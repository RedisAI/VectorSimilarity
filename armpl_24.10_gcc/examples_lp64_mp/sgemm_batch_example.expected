ARMPL example for SGEMM_BATCH
-------------------------------------------------------

******** Group:  1
**** Matrix A
** Trans option: N
     2.000E+00    3.000E+00    4.000E+00
     3.000E+00    4.000E+00    5.000E+00

**** Matrix B
** Trans option: N
     0.000E+00   -1.000E+00
     1.000E+00    0.000E+00
     2.000E+00    1.000E+00

** alpha :   1.000
** beta  :   0.000

**** Matrix C
     1.100E+01    2.000E+00
     1.400E+01    2.000E+00

=============end entry
**** Matrix A
** Trans option: N
     4.000E+00    5.000E+00    6.000E+00
     5.000E+00    6.000E+00    7.000E+00

**** Matrix B
** Trans option: N
     3.000E+00    2.000E+00
     4.000E+00    3.000E+00
     5.000E+00    4.000E+00

** alpha :   1.000
** beta  :   0.000

**** Matrix C
     6.200E+01    4.700E+01
     7.400E+01    5.600E+01

=============end entry
=============end group=============
******** Group:  2
**** Matrix A
** Trans option: T
     6.000E+00    7.000E+00
     7.000E+00    8.000E+00

**** Matrix B
** Trans option: T
     6.000E+00    5.000E+00
     7.000E+00    6.000E+00
     8.000E+00    7.000E+00

** alpha :   1.000
** beta  :   0.000

**** Matrix C
     7.100E+01    8.400E+01    9.700E+01
     8.200E+01    9.700E+01    1.120E+02

=============end entry
**** Matrix A
** Trans option: T
     8.000E+00    9.000E+00
     9.000E+00    1.000E+01

**** Matrix B
** Trans option: T
     9.000E+00    8.000E+00
     1.000E+01    9.000E+00
     1.100E+01    1.000E+01

** alpha :   1.000
** beta  :   0.000

**** Matrix C
     1.440E+02    1.610E+02    1.780E+02
     1.610E+02    1.800E+02    1.990E+02

=============end entry
**** Matrix A
** Trans option: T
     1.000E+01    1.100E+01
     1.100E+01    1.200E+01

**** Matrix B
** Trans option: T
     1.200E+01    1.100E+01
     1.300E+01    1.200E+01
     1.400E+01    1.300E+01

** alpha :   1.000
** beta  :   0.000

**** Matrix C
     2.410E+02    2.620E+02    2.830E+02
     2.640E+02    2.870E+02    3.100E+02

=============end entry
=============end group=============
******** Group:  3
**** Matrix A
** Trans option: N
     1.200E+01    1.300E+01    1.400E+01
     1.300E+01    1.400E+01    1.500E+01
     1.400E+01    1.500E+01    1.600E+01

**** Matrix B
** Trans option: N
     1.500E+01    1.400E+01
     1.600E+01    1.500E+01
     1.700E+01    1.600E+01

** alpha :   1.000
** beta  :   0.000

**** Matrix C
     6.260E+02    5.870E+02
     6.740E+02    6.320E+02
     7.220E+02    6.770E+02

=============end entry
**** Matrix A
** Trans option: N
     1.500E+01    1.600E+01    1.700E+01
     1.600E+01    1.700E+01    1.800E+01
     1.700E+01    1.800E+01    1.900E+01

**** Matrix B
** Trans option: N
     1.800E+01    1.700E+01
     1.900E+01    1.800E+01
     2.000E+01    1.900E+01

** alpha :   1.000
** beta  :   0.000

**** Matrix C
     9.140E+02    8.660E+02
     9.710E+02    9.200E+02
     1.028E+03    9.740E+02

=============end entry
**** Matrix A
** Trans option: N
     1.800E+01    1.900E+01    2.000E+01
     1.900E+01    2.000E+01    2.100E+01
     2.000E+01    2.100E+01    2.200E+01

**** Matrix B
** Trans option: N
     2.100E+01    2.000E+01
     2.200E+01    2.100E+01
     2.300E+01    2.200E+01

** alpha :   1.000
** beta  :   0.000

**** Matrix C
     1.256E+03    1.199E+03
     1.322E+03    1.262E+03
     1.388E+03    1.325E+03

=============end entry
**** Matrix A
** Trans option: N
     2.100E+01    2.200E+01    2.300E+01
     2.200E+01    2.300E+01    2.400E+01
     2.300E+01    2.400E+01    2.500E+01

**** Matrix B
** Trans option: N
     2.400E+01    2.300E+01
     2.500E+01    2.400E+01
     2.600E+01    2.500E+01

** alpha :   1.000
** beta  :   0.000

**** Matrix C
     1.652E+03    1.586E+03
     1.727E+03    1.658E+03
     1.802E+03    1.730E+03

=============end entry
=============end group=============
